### PR TITLE
freecad: add freecad-wayland to fix native wayland

### PR DESCRIPTION
**Pull request by ppenguin**
## Description of changes

The current `FreeCAD` in `nixpkgs` would only run on `xwayland` when running in a `wayland` environment.
This PR adds a `freecad-wayland` attribute that runs natively on `wayland` (i.e. no `xwayland`). It is for now done under a separate attribute to minimise the impact on the existing package (since the wrapper sets the `QPA_PLATFORM` on the original package).

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by AndersonTorres** [comment]
@ofborg build freecad-wayland

**Comment by AndersonTorres** [comment]
Nothing has passed on ofborg.

**Comment by ppenguin** [comment]
> Nothing has passed on ofborg.

Was it hanging?
It looks like now all has passed (or skipped)

**Comment by ppenguin** [comment]
Ah, now it seems green, so somebody with write access could merge this?
Cheers!

File: pkgs/applications/graphics/freecad/default.nix [hunk 1]
```diff
@@ -140,7 +143,7 @@ stdenv.mkDerivation (finalAttrs: {
   qtWrapperArgs = [
     "--set COIN_GL_NO_CURRENT_CONTEXT_CHECK 1"
     "--prefix PATH : ${libredwg}/bin"
-    "--set QT_QPA_PLATFORM xcb"
+    (lib.optional (!withWayland) "--set QT_QPA_PLATFORM xcb")
   ];
```
Comment (AndersonTorres) [inline pkgs/applications/graphics/freecad/default.nix:147]: Nitty:

```suggestion
  ] ++ lib.optionals (!withWayland) [ "--set QT_QPA_PLATFORM xcb" ];
```

**Comment by AndersonTorres** [comment]
Squash the commits.

**Comment by ppenguin** [comment]
> Squash the commits.

done

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -31967,6 +31967,8 @@ with pkgs;
       shiboken2;
   };
 
+  freecad-wayland = pkgs.freecad.override { withWayland = true; };
```
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:31970]:
```suggestion
  freecad-wayland = freecad.override { withWayland = true; };
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (ppenguin) [reply pkgs/top-level/all-packages.nix:31970]: ah yes, of course...

**Review by AndersonTorres — APPROVED** [review]

**Review by MaxwellDupre — APPROVED** [review]
LGTM

**Comment by ppenguin** [comment]
> Fix the conflicts

Done (finally)

**Comment by AndersonTorres** [comment]
Why the so huge diff in all-packages.nix?

File: pkgs/by-name/fr/freecad/package.nix [hunk 3]
```diff
@@ -54,144 +58,151 @@ let
...
+        dimensions or extract design details from 3D models to create high quality
+        production ready drawings.
+
+        FreeCAD is designed to fit a wide range of uses including product design,
+        mechanical engineering and architecture. Whether you are a hobbyist, a
+        programmer, an experienced CAD user, a student or a teacher, you will feel
+        right at home with FreeCAD.
+      '';
+      license = lib.licenses.lgpl2Plus;
+      maintainers = with lib.maintainers; [viric gebner AndersonTorres];
```
Comment (AndersonTorres) [inline pkgs/by-name/fr/freecad/package.nix:205]: What formatter are you using?

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Please undo the formatting changes in this PR. We want to keep the previous formatting done with the nixfmt rfc most likely.

File: pkgs/by-name/fr/freecad/package.nix (hunk 3 above)
Comment (ppenguin) [reply pkgs/by-name/fr/freecad/package.nix:205]: Oops, sorry I missed that one, that was unintended! It appears to be the default formatter that `nvf` is using... I'll reconfig and revert that part...

**Comment by ppenguin** [comment]
> Please undo the formatting changes in this PR. We want to keep the previous formatting done with the nixfmt rfc most likely.

Done (was unintended, caused by a wrong nvf/nvim default setting).

**Review by AndersonTorres — APPROVED** [review]

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**